### DISCUSSION TITLE
Formulating an Exam Schema

### DISCUSSION THREAD

**Discussion by charliemahana** [UTDNebula/nebula-api#52] [Ideas]
# Formulating an Exam Schema
This GitHub discussion is meant to serve as a thread specifically for documenting the design of the exam schema.

**Comment by charliemahana** [comment]
In order to best represent credit bearing exams at UT Dallas, I suggest that we create an abstract Exam class to represent all possible credit bearing exams and then create subclasses for each distinct exam type such as AP, IB, and CLEP. This would enable us to take to take advantage of polymorphism where applicable while enabling us to better represent the individual exam types.

**Comment by jpahm** [comment]
I think this is a good idea. Do you have any specific ideas for what the abstract class should require as far as properties go?

As far as I know, the properties shared between these tests (that are actually useful to us, anyways) would be as follows:

- Date test was taken
- Score received
- Maximum score
- Passing score (i.e. score required for credit to be applied)
- Subtype (i.e. for AP tests, a subtype could be "Macroeconomics")

Let me know of your thoughts.

**Reply by charliemahana** [reply]
The only thing that I can really think of that applies to every single test would be a name property. I would implement AP, IB, and CLEP tests as derived subclasses each handling their own specifics. For example, I would map each score on an AP test to the courses it yields credit for at UTD. For example, a score of 3 yields credit for HIST 1301 whereas a 4 or 5 yield credit for HIST 1301 and HIST 1302. I'm not too familiar with IB or CLEP exams so I can't provide a similar comparison for those exams but the principle should still hold.

**Reply by jpahm** [reply]
I agree on the point of having a name as an abstract property, and I think I understand better now what this is trying to accomplish. I was initially under the impression that we were trying to model exam scores, not necessarily the exams *themselves*. That being said, I think this implementation below fits what you've described:
```typescript
class Course {
	// Course implementation here....
}

abstract class Exam {
	Name: string = "";
};

type ScoreYields = {
	[n: number]: Course[] | null
};

class APExam extends Exam {
	Yields: ScoreYields = {};
	constructor(name: string, yields: ScoreYields) {
		super();
		this.Name = name;
		this.Yields = yields;
	}
};

let Hist1301: Course = new Course();
let Hist1302: Course = new Course();

let HistExam: APExam = new APExam("History", {
	0: null,
	1: null,
	2: null,
	3: [Hist1301],
	4: [Hist1301, Hist1302],
	5: [Hist1301, Hist1302]
});
```
Let me know what you think. It's not a perfect implementation, but a good baseline for discussion.

**Reply by jmyrick02** [reply]
I think your AP Exam implementation works well generally but if we want full coverage we need to handle the subscore edge case in the case of AP Calc BC (AB subscore). For that, it may be a good idea to add a subscore field that has its own ScoreYields object only for use on AP Calc tests. In addition, we must be sure to represent SCH free electives in some form (perhaps as a dummy course with only SCH as the school and the number of credit hours).

In addition, certain majors are restricted from applying AP credits or have certain exceptions. For example, when I was claiming AP credit, my advisor told me that cs majors can't get any credit from AP Statistics, not even free SCH credits despite what's stated on the AP credit site (I cannot find any reference to this in official materials). In addition, Physics majors have a PHYS 3416 edge case as pointed out by Charlie ("Physics Majors: Credit for 2326 + 2126 is applied only if PHYS 3416 is passed with a grade of C or higher; otherwise you earn 4 SCH of free electives. Other Majors: Credit for PHYS 2326 + 2126 with a score of 4 or 5.")

Handling the major-specific edge cases may be somewhat difficult to implement elegantly, especially considering some info on major-specific AP credit policies may only be available from advisors and not publicly available and the PHYS 3416 edge case is somewhat complex as it depends on a class you take during college beyond the test.

**Reply by charliemahana** [reply]
Since college board says that the AP Calc BC subscore should be treated as an AP Calc AB score, I don't think we actually need to worry about adding more data for an AP exam. We just treat subscores as separate exams. https://apstudents.collegeboard.org/about-ap-scores/special-score-structure-calculus-bc

**Reply by jmyrick02** [reply]
I wrote a schema of the most common exams I could think of at UT Dallas. These are either credit-bearing or placement tests for prerequisites. The `yields` field only gives the credit received, which is why the ALEKS exam has an empty yields dictionary, since it only offers placement, not credit. This does not handle that PHYS 3416 edge case, so input on how to address that would be appreciated.

```ts
ExamType = "AP" | "ALEKS" | "CLEP" | "IB" | "CS placement"
  
Exam = {
  "_id": ObjectId,
  "type": ExamType,
  "yields": Dictionary<int, Array<ObjectId>>,
}
  
APExam extends Exam = {
  "_id": ObjectId,
  "type": "AP",
  "name": string,
  "yields": Dictionary<int, Array<ObjectId>>,
}
  
ALEKSExam extends Exam = {
  "_id": ObjectId,
  "type": "ALEKS",
  "yields": {},
}
  
CLEPExam extends Exam = {
  "_id": ObjectId,
  "type": "CLEP",
  "name": string,
  "yields": Dictionary<int, Array<ObjectId>>,
}
 
IBExam extends Exam = {
  "_id": ObjectId,
  "type": "IB",
  "name": string,
  "level": string,
  "yields": Dictionary<int, Array<ObjectId>>,
}
  
CSPlacementExam extends Exam = {
  "_id": ObjectId,
  "type": "CS placement",
  "yields": {70: [CS1336._id, CS1136._id]},
}
```
